json-everything.net/wwwroot/md/vocabs-data-2022.md

```diff
@@ -16,66 +16,76 @@ The intent for this keyword is to cover the use cases discussed across several i
+- Keys that are ignored by the host schema MUST also be ignored by the formed schema.
 
-Implementations SHOULD validate the formed schema against the containing schema's meta-schema (as specified by `$schema`) to ensure that it is a syntactically valid schema object.
+Implementations SHOULD validate the formed schema against the host schema's meta-schema (as specified by `$schema`) to ensure that it is a syntactically valid schema object.
 
-### 3.3 URI Resolution
+### 4.3 URI Resolution
 
 The values of `data` are dereferenced in different ways depending on the format of the value.
 
 If the value is a JSON Pointer, it is resolved against the instance root.
 
 If the value is a Relative JSON Pointer, it is resolved against the instance at the location currently being evaluated.
 
-If the value is a URI reference, it must be resolved in accordance with the rules of `$ref` resolution for the relevant JSON Schema specification (e.g. [draft 2020-12, §8.2](https://json-schema.org/draft/2020-12/json-schema-core.html#name-base-uri-anchors-and-derefe)).  However, unlike `$ref` which requires that the indicated data must represent a valid schema object, a `data` reference can identify any value which is valid for the associated keyword.
+Otherwise, it must be resolved in accordance with the rules of `$ref` resolution for the relevant JSON Schema specification (e.g. [draft 2020-12, §8.2](https://json-schema.org/draft/2020-12/json-schema-core.html#name-base-uri-anchors-and-derefe)).  However, unlike `$ref` which requires that the indicated data must represent a valid schema object, a `data` reference can identify any value which is valid for the associated keyword.
 
-Implementations SHOULD provide a means to pre-load and cache any external reference but MAY be configured to fetch external documents at evaluation time.  Documents fetched from URIs which contain a JSON Pointer fragment MUST be interpreted using a media type that allows resolution of such fragments.
+Because JSON Pointers and Relative JSON Pointers are syntactically valid URIs, the value MUST be checked in the sequence indicated above in order to properly identify the pointer types from other URIs.
 
-If any single reference cannot be resolved, validation MUST fail; otherwise the full value at the specified location is returned.
+For each successfully resolved reference, the full value at the specified location MUST be returned.
 
-If the resolved value is not valid for the associated keyword, validation MUST fail.
+If a reference cannot be resolved, or if a resolved value is not valid for the associated keyword, evaluation MUST halt.  Implementations SHOULD use native features of their language to report the failure as appropriate.  Implementations MAY continue to attempt to resolve other references so that multiple resolution failures can be reported together, however further schema evaluation MUST NOT continue.
```

Comment:
Interesting - I had not entirely put this together with the `$ref` stuff.

I do think I agree with it.  Were this purely about instance references, you could classify the failure as a validation failure (e.g. this keyword requires two parts of the instance to exist in some sort of relation, and since they don't that is a validation failure because it's all about the instance).  But the external data aspect means that unresolved references are errors, not validation failures.

Reply:
Yeah, my thinking is that the "unresolved reference" aspect trumps the "in the instance" aspect, especially because not all references are in the instance and all dereference attempts should be treated the same.